### PR TITLE
RSE-196: Create Convert to Pledge/Contribution and "View Pledge/Contribution" buttons

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR, the "Convert to Pledge/Contribution" and "View Pledge/Contribution" buttons has been added as Case Actions.

## After
after

### New Angular Module
A new Angular module `prospect` has been created, which serves all Angular JS functionality for CiviProspect extension.

### Case Actions Added via JS
Inside the Angular JS code, a new file `app.config.js` is created, and in the file the new case actions like "Convert to Pledge/Contribution" and "View Pledge/Contribution" has been added.

### Prospecting Constant
Inside `app.constants.js` file, `ProspectGlobalValues` constant is created which serves all constants for the app.

### New Services Created for the New Case Actions
Two new services has been created,
  1. `ConvertProspectCaseAction`
  2. `ViewPledgeContributionCaseAction`

They follow the same architecture as defined in https://github.com/compucorp/uk.co.compucorp.civicase/pull/228

### New API Created
`ProspectConverted.getpaymentinfo`, api is created.
It exposes the existing `CRM_Prospect_BAO_ProspectConverted::getPaymentInfo()` via API.

### Others
1. Removed Old Unused Files. Like `Prospect.Page.CaseView.js`, `templates/CRM/Case/Page/Tab.extra.tpl` and `_prospect_civicrm_alterTemplateFile_CRM_Case_Page_Tab` function.
2. Semistandard JS linter is added.

### CODE REVIEW

File: ang/prospect/app.config.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
+(function (angular, $, _) {
+  var module = angular.module('prospect');
+
+  module.run(function () {
+    (function init () {
+      CRM.civicase.caseActions = CRM.civicase.caseActions.concat(getCaseActionsForProspect());
```
Comment (reneolivo) [inline ang/prospect/app.config.js]: This will add the actions to the end of the list, is that ok?

File: ang/prospect/services/convert-prospect-case-action.service.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    };
+
+    /**
+     * Checks if the Case is Converted to Prospect
+     *
+     * @param {Array} cases
+     * @return {Promise}
+     */
+    function checkIfConvertedToProspect (cases) {
+      if (cases[0]) {
```
Comment (reneolivo) [inline ang/prospect/services/convert-prospect-case-action.service.js]:
```js
if (!cases[0]) {
  return;
}

// ...
```

File: ang/prospect/services/convert-prospect-case-action.service.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        query: {
+          action: 'add',
+          reset: 1,
+          cid: contactID,
+          context: action.type,
+          caseid: caseId
+        }
+      };
+
+      return popupPath;
```
Comment (reneolivo) [inline ang/prospect/services/convert-prospect-case-action.service.js]:
```
return {
  // ...
};
```

Also, why is this called `doAction` if it only returns some popup configuration? All actions return configurations? If so, just call the method `getPopupParams` or similar.

File: ang/prospect/services/convert-prospect-case-action.service.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+     * @param {Array} cases
+     * @return {Promise}
+     */
+    function checkIfConvertedToProspect (cases) {
+      if (cases[0]) {
+        var caseID = cases[0].id;
+
+        return crmApi('ProspectConverted', 'get', {
+          'sequential': 1,
+          'prospect_case_id': caseID
```
Comment (reneolivo) [inline ang/prospect/services/convert-prospect-case-action.service.js]: We don't need to add single quotes `'...'` to these properties. Semistandard doesn't warn against this?

File: ang/prospect/services/convert-prospect-case-action.service.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * Click event handler for the Action
+     *
+     * @param {Array} cases
+     * @param {Object} action
+     * @param {Function} callbackFn
+     */
+    this.doAction = function (cases, action, callbackFn) {
+      if (action.type !== 'contribution' && action.type !== 'pledge') {
```
Comment (reneolivo) [inline ang/prospect/services/convert-prospect-case-action.service.js]:
```js
if (_.includes(['contribution', 'pledge'], action.type)) {
  // ...
}
```

Also, why is this here and not on `isActionAllowed` ? when I run a method I expect that method to work and not exit silently with a `return;`.

File: ang/prospect/services/convert-prospect-case-action.service.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    /**
+     * Check if Case Type Category is Prospecting.
+     *
+     * @return {Boolean}
+     */
+    function checkIfProspectingCaseTypeCategory () {
+      var filtersQueryParams = $location.search().cf;
+      if (filtersQueryParams) {
```
Comment (reneolivo) [inline ang/prospect/services/convert-prospect-case-action.service.js]: I'd put a space between the var block and the if block.

Comment (reneolivo) [reply ang/prospect/services/convert-prospect-case-action.service.js]: Also:

```js
function checkIfProspectingCaseTypeCategory () {
  var filtersQueryParams = $location.search().cf;

  // exit early:
  if (!filtersQueryParams) {
    return false; // don't forget to return false instead of undefined.
  }

  // ...
}
```

File: ang/prospect/services/view-pledge-contribution-case-action.service.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+        checkIfConvertedToProspect(caseID)
+          .then(function () {
+            getPaymentInfo(caseID);
+          });
+      }
+    };
+
+    /**
+     * Checks if the Action is allowed
```
Comment (reneolivo) [inline ang/prospect/services/view-pledge-contribution-case-action.service.js]: it will also be good to explain how the action is allowed.

File: ang/prospect/services/view-pledge-contribution-case-action.service.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+     *
+     * @param {Object} action
+     * @param {Array} cases
+     * @return {boolean}
+     */
+    this.isActionAllowed = function (action, cases) {
+      var isActionAllowed = action.type === 'pledge'
+        ? prospectConverted.paymentInfo.payment_entity === 'pledge'
+        : prospectConverted.paymentInfo.payment_entity === 'contribute';
```
Comment (reneolivo) [inline ang/prospect/services/view-pledge-contribution-case-action.service.js]: isn't this the same:

```js
var isActionAllowed = action.type === prospectConverted.paymentInfo.payment_entity;
```

File: ang/prospect/services/view-pledge-contribution-case-action.service.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    /**
+     * Click event handler for the Action
+     *
+     * @param {Array} cases
+     * @param {Object} action
+     * @param {Function} callbackFn
+     */
+    this.doAction = function (cases, action, callbackFn) {
+      var contactID = cases[0].client[0].contact_id;
+      var url = prospectConverted.paymentInfo.payment_url + '&cid=' + contactID;
```
Comment (reneolivo) [inline ang/prospect/services/view-pledge-contribution-case-action.service.js]: return directly.

File: ang/prospect/services/view-pledge-contribution-case-action.service.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+     * @param {String/Int} caseID
+     */
+    function getPaymentInfo (caseID) {
+      if (prospectConverted.isConverted) {
+        crmApi('ProspectConverted', 'getpaymentinfo', {
+          'prospect_case_id': caseID
+        }).then(function (paymentInfo) {
+          prospectConverted.paymentInfo = paymentInfo;
+        });
+      }
```
Comment (reneolivo) [inline ang/prospect/services/view-pledge-contribution-case-action.service.js]: why do we need an `if` here? I find some issues with this function in general:

- it's called `getPaymentInfo`, but doesn't return anything.
- it might or might not "get" what I want depending on a condition I'm not aware of (`if (prospectConverted.isConverted) {`. What I mean is, anyone calling `getPaymentInfo` will not know why they are not getting the payment info if `prospectConverted.isConverted` is set to false.
- It's overwriting a global value `prospectConverted.paymentInfo`, so it's a setter, not a getter.

File: ang/prospect/services/view-pledge-contribution-case-action.service.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    function checkIfProspectingCaseTypeCategory () {
+      var filtersQueryParams = $location.search().cf;
+      if (filtersQueryParams) {
+        var caseTypeCategory = JSON.parse(filtersQueryParams).case_type_category;
+
+        return caseTypeCategory === ProspectGlobalValues.caseTypeCategory;
+      }
+    }
+  }
+})(angular, CRM.$, CRM._);
```
Comment (reneolivo) [inline ang/prospect/services/view-pledge-contribution-case-action.service.js]: After checking all the code and seeing how `checkIfConvertedToProspect` and `checkIfProspectingCaseTypeCategory` I'd try to put those intro a service.

`checkIfProspectingCaseTypeCategory` should be on a service of its own, but I think it needs a bit of rework because it's not clear it does this check from the URL params.

I'd have a service like this instead:

```js
function getCaseTypeCategoryFromUrlParams() {
  var filtersQueryParams = $location.search().cf;

  if (!filtersQueryParams) {
    return;
  }

  return JSON.parse(filtersQueryParams).case_type_category
}
```

And use this service like this:

```js
getCaseTypeCategoryFromUrlParams() === ProspectGlobalValues.caseTypeCategory;
```

Comment (reneolivo) [reply ang/prospect/services/view-pledge-contribution-case-action.service.js]: for `checkIfConvertedToProspect` I'd refactor it like this:

```js
function ProspectConvertedService() {
  this.getProspectIsConverted = getProspectIsConverted;
  this. getPaymentInfo = getPaymentInfo;

  function getProspectIsConverted(caseId) {
     return crmApi('ProspectConverted', 'get', {
        sequential: 1,
        prospect_case_id: caseId
      }).then(function (caseData) {
        return caseData.count > 0;
      });
  }

  function getPaymentInfo (caseId) {
    crmApi('ProspectConverted', 'getpaymentinfo', {
      prospect_case_id: caseId
    }).then(function (paymentInfo) {
      return paymentInfo;
    });
  }
}
```

Then on the actual action:

```js
const isProspectConverted = false;
const prospectConvertedPaymentInfo = {};

// ...

this.refreshData = function (cases) {
  if (!cases[0]) {
    return;
  }

  ProspectConvertedService.getProspectIsConverted(cases[0].id)
    .then(function (isConverted) {
      isProspectConverted = isConverted;

      return isConverted
        ?  ProspectConvertedService.getPaymentInfo(cases[0].id)
        :  {};
    })
    .then(function (paymentInfo) {
      prospectConvertedPaymentInfo = paymentInfo;
    });
}
```

File: ang/prospect/app.config.js (hunk 1 above)
Comment (deb1990) [reply ang/prospect/app.config.js]: There was not any clear instruction on where to end the new actions. But I changed it to appear before others 👍

File: ang/prospect/services/convert-prospect-case-action.service.js (hunk 3 above)
Comment (deb1990) [reply ang/prospect/services/convert-prospect-case-action.service.js]: For now all services are not consistent, some do the action, and some return a url. But this will be fixed in future
https://github.com/compucorp/uk.co.compucorp.civicase/pull/228#discussion_r309537750

So keeping `doAction` for now, because thats the name across all services.

File: ang/prospect/services/convert-prospect-case-action.service.js (hunk 4 above)
Comment (deb1990) [reply ang/prospect/services/convert-prospect-case-action.service.js]: Oops, I copied from API explorer, but no Semistandard does not complain.

File: ang/prospect/services/convert-prospect-case-action.service.js (hunk 5 above)
Comment (deb1990) [reply ang/prospect/services/convert-prospect-case-action.service.js]: It was copied from existing code. Fixed.

File: ang/prospect/services/view-pledge-contribution-case-action.service.js (hunk 8 above)
Comment (deb1990) [reply ang/prospect/services/view-pledge-contribution-case-action.service.js]: No, because for contribution,
action type is `contribution` and payment_entity is `contribute`. But changed the logic to make it easier to understand

```js
var actionTypeMapping = {
  pledge: 'pledge', contribution: 'contribute'
};
var isActionAllowed = actionTypeMapping[action.type] === prospectConverted.paymentInfo.payment_entity;
```

File: ang/prospect/services/view-pledge-contribution-case-action.service.js (hunk 10 above)
Comment (deb1990) [reply ang/prospect/services/view-pledge-contribution-case-action.service.js]: Moved the setter functionality and the if condition outside.

File: ang/prospect/services/view-pledge-contribution-case-action.service.js (hunk 11 above)
Comment (deb1990) [reply ang/prospect/services/view-pledge-contribution-case-action.service.js]: Good tip. Fixed.

File: ang/prospect/app.config.js (hunk 1 above)
Comment (reneolivo) [reply ang/prospect/app.config.js]: I was just asking :3

**Review by reneolivo — APPROVED** [review]
